File: java/dev/mlm/retorrent/TorrentService.java
package dev.mlm.retorrent;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.IBinder;
import java.io.InputStream;
import java.io.ByteArrayOutputStream;

public class TorrentService extends Service {
    static {
        System.loadLibrary("retorrent");
    }
    public void onCreate() {
        super.onCreate();
        nativeOnCreate(this);
    }
    public int onStartCommand(Intent i, int f, int sid) {
        if (i != null && Intent.ACTION_VIEW.equals(i.getAction())) {
            Uri uri = i.getData();
            if (uri != null) {
                try {
                    InputStream is = getContentResolver().openInputStream(uri);
                    if (is != null) {
                        ByteArrayOutputStream baos = new ByteArrayOutputStream();
                        byte[] buf = new byte[4096];
                        int n;
                        while ((n = is.read(buf)) >= 0) {
                            baos.write(buf, 0, n);
                        }
                        is.close();
                        nativeOnTorrentData(baos.toByteArray());
                    }
                } catch (Exception e) {
                    // ignore
                }
            }
        }
        return START_STICKY;
    }
    public IBinder onBind(Intent i) {
        return null;
    }
    public void onDestroy() {
        nativeOnDestroy();
        super.onDestroy();
    }
    private static native void nativeOnCreate(Context context);
    private static native void nativeOnDestroy();
    private static native void nativeOnTorrentData(byte[] data);
}

File: java/dev/mlm/retorrent/TorrentService.kt
package org.mlm.retorrent

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.util.Log

class TorrentService : Service() {
    override fun onCreate() {
        super.onCreate()
        nativeOnCreate(this)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent?.action == Intent.ACTION_VIEW) {
            intent.data?.let { uri ->
                try {
                    val bytes = readTorrentBytes(uri) ?: return START_NOT_STICKY
                    nativeOnTorrentData(bytes)
                } catch (e: Exception) {
                    Log.e(TAG, "onStartCommand: failed to read torrent data", e)
                }
            }
        }
        return START_NOT_STICKY
    }

    override fun onTimeout(startId: Int, fgsType: Int) {
        Log.w(TAG, "onTimeout: foreground service timeout (startId=$startId, fgsType=$fgsType)")
        try {
            nativeOnForegroundServiceTimeout()
        } catch (e: Throwable) {
            Log.e(TAG, "onTimeout: native call failed", e)
        }
        stopSelf(startId)
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onDestroy() {
        nativeOnDestroy()
        super.onDestroy()
    }

    private fun readTorrentBytes(uri: android.net.Uri): ByteArray? {
        val maxBytes = 32 * 1024 * 1024
        return contentResolver.openInputStream(uri)?.use { input ->
            val out = java.io.ByteArrayOutputStream()
            val buf = ByteArray(16 * 1024)
            var total = 0

            while (true) {
                val n = input.read(buf)
                if (n < 0) break
                total += n
                if (total > maxBytes) {
                    throw IllegalArgumentException(
                        "Torrent file too large (max ${maxBytes / (1024 * 1024)} MiB)"
                    )
                }
                out.write(buf, 0, n)
            }

            out.toByteArray()
        }
    }

    companion object {
        private const val TAG = "TorrentService"

        init {
            System.loadLibrary("retorrent")
        }

        @JvmStatic private external fun nativeOnCreate(context: Context)
        @JvmStatic private external fun nativeOnDestroy()
        @JvmStatic private external fun nativeOnTorrentData(data: ByteArray)
        @JvmStatic private external fun nativeOnForegroundServiceTimeout()
    }
}
